File: src/python/philosopher.py
import enum
import time
import random
import threading

class StatesPhilosopher(enum.Enum):
    '''
    Representa los posibles estados de los filosofos.
    '''
    thinking = 1
    hungry = 2
    eating = 3
    left_fork = 4
    right_fork = 5
    finished_eat = 6

class Philosopher:

    def __init__(self, count_philosopher=5, max_failed=3):
        self.count_philosopher = count_philosopher
        self.max_failed = max_failed

        self.forks = [ threading.Semaphore(1) for _ in range(self.count_philosopher) ]
        self.used_fork = [ False for _ in range(self.count_philosopher) ]
        self.state_philosopher = [ StatesPhilosopher.thinking 
                                    for _ in range(self.count_philosopher) ]
        self.start_threads()

    def get_left_fork(self, id_philosopher):
        '''
        Devuelve el tenedor izquierdo del filosofo especificado.
        '''
        return id_philosopher
    
    def get_right_fork(self, id_philosopher):
        '''
        Devuelve el tenedor derecho del filosofo especificado.
        '''
        return (id_philosopher + 1) % self.count_philosopher

    def take_left_fork(self, id_philosopher):
        '''
        Intenta tomar el tenedor izquierdo del filosofo especificado.
        '''
        id_left_fork = self.get_left_fork(id_philosopher)
        if not self.used_fork[id_left_fork]:
            self.forks[id_left_fork].acquire()
            self.used_fork[id_left_fork] = True
            print(f'[<=] Filosofo {id_philosopher} cogio el tenedor izquierdo.')
            self.state_philosopher[id_left_fork] = StatesPhilosopher.left_fork

    def take_right_fork(self, id_philosopher):
        '''
        Intenta tomar el tenedor derecho del filosofo especificado.
        '''
        id_right_fork = self.get_right_fork(id_philosopher)
        if not self.used_fork[id_right_fork]:
            self.forks[id_right_fork].acquire()
            self.used_fork[id_right_fork] = True
            print(f'[=>] Filosofo {id_philosopher} cogio el tenedor derecho.')
            self.state_philosopher[id_right_fork] = StatesPhilosopher.right_fork

    def release_left_fork(self, id_philosopher):
        '''
        Libera el tenedor izquierdo del filosofo especificado.
        '''
        id_left_fork = self.get_left_fork(id_philosopher)
        if self.used_fork[id_left_fork]:
            self.forks[id_left_fork].release()
            self.used_fork[id_left_fork] = False
            print(f'[<-] Filosofo {id_philosopher} solto el tenedor izquierdo.')

    def release_right_fork(self, id_philosopher):
        '''
        Libera el tenedor derecho del filosofo especificado.
        '''
        id_right_fork = self.get_right_fork(id_philosopher)
        if self.used_fork[id_right_fork]:
            self.forks[id_right_fork].release()
            self.used_fork[id_right_fork] = False
            print(f'[<-] Filosofo {id_philosopher} solto el tenedor derecho.')

    def eat(self, id_philosopher):
        '''
        Funcion que va ejecutar cada filosofo o hilo.
        '''
        while True:

            if random.randint(0, 1) == 0:

                self.state_philosopher[id_philosopher] = StatesPhilosopher.hungry
                print(f'[@] Filosofo {id_philosopher} hambiento.')

                if not self.used_fork[self.get_left_fork(id_philosopher)]:
                    self.take_left_fork(id_philosopher)
                    failed = 0
                    while True:
                        if failed == self.max_failed:
                            failed = 0
                            self.release_left_fork(id_philosopher)
                            break
                        elif not self.used_fork[self.get_right_fork(id_philosopher)]:
                            failed = 0
                            self.take_right_fork(id_philosopher)
                            print(f'[+] Filosofo {id_philosopher} comiendo.')
                            self.state_philosopher[id_philosopher] = StatesPhilosopher.eating
                            time.sleep(random.randint(3, 5))
                            self.state_philosopher[id_philosopher] = StatesPhilosopher.finished_eat
                            print(f'[!] Filosofo {id_philosopher} termino de comer.')
                            self.release_left_fork(id_philosopher)
                            self.release_right_fork(id_philosopher)
                            break
                        else:
                            failed += 1
                            print(f'[{failed}] Filosofo {id_philosopher} no pudo coger el tenedor derecho porque estaba ocupado.')
                        time.sleep(random.randint(1, 3))

                    self.state_philosopher[id_philosopher] = StatesPhilosopher.thinking
                    print(f'[*] Filosofo {id_philosopher} pensando.')
                    time.sleep(random.randint(2, 5))
                elif not self.used_fork[self.get_right_fork(id_philosopher)]:
                    self.take_right_fork(id_philosopher)
                    failed = 0
                    while True:
                        if failed == self.max_failed:
                            failed = 0
                            self.release_right_fork(id_philosopher)
                            break
                        elif not self.used_fork[self.get_left_fork(id_philosopher)]:
                            failed = 0
                            self.take_left_fork(id_philosopher)
                            print(f'[+] Filosofo {id_philosopher} comiendo.')
                            self.state_philosopher[id_philosopher] = StatesPhilosopher.eating
                            time.sleep(random.randint(3, 5))
                            self.state_philosopher[id_philosopher] = StatesPhilosopher.finished_eat
                            print(f'[!] Filosofo {id_philosopher} termino de comer.')
                            self.release_left_fork(id_philosopher)
                            self.release_right_fork(id_philosopher)
                            break
                        else:
                            failed += 1
                            print(f'[{failed}] Filosofo {id_philosopher} no pudo coger el tenedor izquierdo porque estaba ocupado.')
                        time.sleep(random.randint(1, 3))

                    self.state_philosopher[id_philosopher] = StatesPhilosopher.thinking
                    print(f'[*] Filosofo {id_philosopher} pensando.')
                    time.sleep(random.randint(2, 5))

            else:
                self.state_philosopher[id_philosopher] = StatesPhilosopher.thinking
                print(f'[*] Filosofo {id_philosopher} pensando.')
                time.sleep(random.randint(2, 5))

            time.sleep(1)

    def start_threads(self):
        '''
        Comienza a correr los hilos correspondiente 
        a cada filosofo.
        '''

        threads = [ threading.Thread(target=self.eat, args=(id,)) 
                        for id in range(self.count_philosopher) ]

        for thread in threads:
            thread.start()

        for thread in threads:
            thread.join()

def main():
    
    Philosopher()

if __name__ == '__main__':
    main()

File: src/c#/philosopher.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Threading;

namespace Philosophers
{
    public enum StatesPhilosopher
    {
        thinking = 0,
        hungry = 1,
        eating = 2,
        left_fork = 3,
        right_fork = 4,
        finished_eat = 5
    }

    public class Philosopher
    {
        private Semaphore[] forks;
        private bool[] used_forks;
        private Thread[] philosopherThread;
        private StatesPhilosopher[] statesPhilosopher;

        public int CountPhilosopher { get; private set; }
        public int MaxFailed { get; set; }

        public Philosopher(int countPhilosopher = 5, int maxFailed = 3)
        {
            forks = new Semaphore[countPhilosopher];
            used_forks = new bool[countPhilosopher];
            philosopherThread = new Thread[countPhilosopher];
            statesPhilosopher = new StatesPhilosopher[countPhilosopher];
            CountPhilosopher = countPhilosopher;
            MaxFailed = maxFailed;

            for (int i = 0; i < countPhilosopher; i++)
            {
                used_forks[i] = false;
                forks[i] = new Semaphore(1, 1);
                statesPhilosopher[i] = StatesPhilosopher.thinking;
            }
        }

        private int GetLeftFork(int id_philosopher)
        {
            return id_philosopher;
        }

        private int GetRightFork(int id_philosopher)
        {
            return (id_philosopher + 1) % CountPhilosopher;
        }

        private void TakeLeftFork(int id_philosopher)
        {
            int id_left_fork = GetLeftFork(id_philosopher);
            if (!used_forks[id_left_fork])
            {
                forks[id_left_fork].WaitOne();
                used_forks[id_left_fork] = true;
                Console.WriteLine($"[<=] Filosofo {id_philosopher} cogio el tenedor izquierdo.");
                statesPhilosopher[id_left_fork] = StatesPhilosopher.left_fork;
            }
        }

        private void TakeRightFork(int id_philosopher)
        {
            int id_right_fork = GetRightFork(id_philosopher);
            if (!used_forks[id_right_fork])
            {
                forks[id_right_fork].WaitOne();
                used_forks[id_right_fork] = true;
                Console.WriteLine($"[=>] Filosofo {id_philosopher} cogio el tenedor derecho.");
                statesPhilosopher[id_right_fork] = StatesPhilosopher.right_fork;
            }
        }

        private void ReleaseLeftFork(int id_philosopher)
        {
            int id_left_fork = GetLeftFork(id_philosopher);
            if (used_forks[id_left_fork])
            {
                forks[id_left_fork].Release();
                used_forks[id_left_fork] = false;
                Console.WriteLine($"[<-] Filosofo {id_philosopher} solto el tenedor izquierdo.");
            }
        }

        private void ReleaseRightFork(int id_philosopher)
        {
            int id_right_fork = GetRightFork(id_philosopher);
            if (used_forks[id_right_fork])
            {
                forks[id_right_fork].Release();
                used_forks[id_right_fork] = false;
                Console.WriteLine($"[->] Filosofo {id_philosopher} solto el tenedor derecho.");
            }
        }

        private void Eat(int id_philosopher)
        {
            Random random = new Random();

            while (true)
            {
                if (random.Next(0, 2) == 0)
                {
                    statesPhilosopher[id_philosopher] = StatesPhilosopher.hungry;
                    Console.WriteLine($"[@] Filosofo {id_philosopher} hambiento.");

                    if (!used_forks[GetLeftFork(id_philosopher)])
                    {
                        TakeLeftFork(id_philosopher);
                        int failed = 0;

                        while (true)
                        {
                            if (failed == MaxFailed)
                            {
                                failed = 0;
                                ReleaseLeftFork(id_philosopher);
                                break;
                            }
                            else if (!used_forks[GetRightFork(id_philosopher)])
                            {
                                failed = 0;
                                TakeRightFork(id_philosopher);
                                Console.WriteLine($"[+] Filosofo {id_philosopher} comiendo.");
                                statesPhilosopher[id_philosopher] = StatesPhilosopher.eating;
                                Thread.Sleep(TimeSpan.FromSeconds(random.Next(3, 6)));
                                statesPhilosopher[id_philosopher] = StatesPhilosopher.finished_eat;
                                Console.WriteLine($"[!] Filosofo {id_philosopher} termino de comer.");
                                ReleaseLeftFork(id_philosopher);
                                ReleaseRightFork(id_philosopher);
                                break;
                            }
                            else
                            {
                                failed += 1;
                                Console.WriteLine($"[{failed}] Filosofo {id_philosopher} no pudo coger el tenedor derecho porque estaba ocupado.");
                            }
                            Thread.Sleep(TimeSpan.FromSeconds(random.Next(1, 4)));
                        }
                        statesPhilosopher[id_philosopher] = StatesPhilosopher.thinking;
                        Console.WriteLine($"[*] Filosofo {id_philosopher} pensando.");
                        Thread.Sleep(TimeSpan.FromSeconds(random.Next(2, 6)));
                    }
                    else if (!used_forks[GetRightFork(id_philosopher)])
                    {
                        TakeRightFork(id_philosopher);
                        int failed = 0;

                        while (true)
                        {
                            if (failed == MaxFailed)
                            {
                                failed = 0;
                                ReleaseRightFork(id_philosopher);
                                break;
                            }
                            else if (!used_forks[GetLeftFork(id_philosopher)])
                            {
                                failed = 0;
                                TakeLeftFork(id_philosopher);
                                Console.WriteLine($"[+] Filosofo {id_philosopher} comiendo.");
                                statesPhilosopher[id_philosopher] = StatesPhilosopher.eating;
                                Thread.Sleep(TimeSpan.FromSeconds(random.Next(3, 6)));
                                statesPhilosopher[id_philosopher] = StatesPhilosopher.finished_eat;
                                Console.WriteLine($"[!] Filosofo {id_philosopher} termino de comer.");
                                ReleaseLeftFork(id_philosopher);
                                ReleaseRightFork(id_philosopher);
                                break;
                            }
                            else
                            {
                                failed += 1;
                                Console.WriteLine($"[{failed}] Filosofo {id_philosopher} no pudo coger el tenedor izquierdo porque estaba ocupado.");
                            }
                            Thread.Sleep(TimeSpan.FromSeconds(random.Next(1, 4)));
                        }
                        statesPhilosopher[id_philosopher] = StatesPhilosopher.thinking;
                        Console.WriteLine($"[*] Filosofo {id_philosopher} pensando.");
                        Thread.Sleep(TimeSpan.FromSeconds(random.Next(2, 6)));
                    }
                    else
                    {
                        statesPhilosopher[id_philosopher] = StatesPhilosopher.thinking;
                        Console.WriteLine($"[*] Filosofo {id_philosopher} pensando.");
                        Thread.Sleep(TimeSpan.FromSeconds(random.Next(2, 6)));
                    }
                    Thread.Sleep(TimeSpan.FromSeconds(1));
                }
            }
        }

        public void Run()
        {
            for (int i = 0; i < philosopherThread.Length; i++)
            {
                int idx = i;
                philosopherThread[idx] = new Thread(() => Eat(idx));
                philosopherThread[idx].Start();
            }

            for (int i = 0; i < philosopherThread.Length; i++)
            {
                philosopherThread[i].Join();
            }
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            Philosopher forForkQueue = new Philosopher(5, 3);
            forForkQueue.Run();
        }
    }
}
